### DISCUSSION TITLE
how to set the sliding-window size of lz4_frame compression?

### DISCUSSION THREAD

**Discussion by xiaodou1117** [lz4/lz4#1534] [Q&A] [answered]
i want to compress a file with the sliding-window size of the file size.
like compress a 8KB file, and the sliding-window size is 8k,too.

**Comment by Cyan4973** [comment]
There is no window size setting in `lz4`,
it's always implicitly 64 KB, or effectively the size of the file when it's smaller (8kb in your example).

**Reply by xiaodou1117** [reply]
thank you for that.
But this is why I come to ask. I was compressing a 128KB file all together using `LZ4F_compressFrame()`. I set the `LZ4F_blockSizeID_t` as `LZ4F_max64KB`  ,then I set it as `LZ4F_max256KB`  but the ratio did not change. 
So what  actually the `LZ4F_blockSizeID_t` is as it is not the window size？
And if I want to compress the 128KB file all together，what should I do？（maybe I should refer to the framecompress.c ?)
I‘m new to lz4, sorry for too many questions.

**Reply by Cyan4973** [reply]
This setting defines the **block** size. A frame can contain multiple blocks.
If you select `LZ4F_max64KB`, it will need 2 blocks to compress 128 KB.
If you select `LZ4F_max256KB` or anything higher, it will need only one block.

I would expect the `LZ4F_max64KB` variant to be slightly less efficient, because of headers for 2 blocks, but really only very slightly, so it probably doesn't matter.

**Reply by xiaodou1117** [reply]
so, if I chose `LZ4F_max256KB`, I can compress 128KB in one block,and this don't need me to frame the file like using `LZ4F_compressBegin()` and `LZ4F_compressUpdate()` but only `LZ4F_compressFrame()` to compress the whole block. And I can also get the whole dependence of the 128KB file.

**Reply by xiaodou1117** [reply]
and if I select `LZ4F_max64KB`, can block2 compress with the dependence of block 1 just using `LZ4F_compressFrame()`？

**Reply by Cyan4973** [reply, marked as answer]
You can always use `LZ4F_compressFrame()`.
As long as you have the entire input available at the moment you call the compression function, `LZ4F_compressFrame()` is fine.
The fact that it will break down data internally into 1 or 2 blocks is more an implementation detail.

`LZ4F_compressBegin()` and `LZ4F_compressUpdate()` are required when you don't have the whole input available at the first call, and need to stream it progressively into the compression state. It's a more advanced use case.

**Reply by xiaodou1117** [reply]
I get it. Thank you for your patient answer

**Reply by xiaodou1117** [reply]
maybe one last question：
if I compress 128KB file and the window size is implicitly 64 KB，dose this mean I can‘t compress block2 with block1 by just using LZ4F_compressFrame()？

**Reply by Cyan4973** [reply]
`LZ4F_compressFrame()` works, irrespective of the nb of internal blocks created.
Just try it, you'll observe the outcome.

**Reply by xiaodou1117** [reply]
thank your for your patient answer！
